Improve Code Quality

bhgomes opened the pull request:
This PR is based on the `edition` branch https://github.com/rust-zkp/ark-plonk/pull/37

## Changes
- fixed `clippy` suggestions or marked where `clippy` is wrong
- fixed broken doc links
- removed `get_domain_attrs` in favor of `EvaluationDomainExt`
- removed unnecessary import renaming
- general best-practices improvements

davidnevadoc committed:
Add logic operations between field elements

Add explicit loop for logic operations AND and XOR in feature
`trace` code.

davidnevadoc committed:
Update to edition 2021

bhgomes committed:
fix: improve code quality standards across the crate

- fixed clippy suggestions or marked where clippy is wrong
- fixed broken doc links
- removed `get_domain_attrs` in favor of `EvaluationDomainExt`
- general best-practices improvements

bhgomes committed:
fix: remove unnecessary import renaming

bhgomes committed:
fix: use Range instead of RangeInclusive

bhgomes committed:
fix: improve CI workflow, include clippy and docs

bhgomes committed:
chore: fix typo

davidnevadoc commented on src/proof_system/proof.rs:
```diff
@@ -491,11 +493,10 @@ fn compute_barycentric_eval<F: PrimeField>(
-    let group_gen_inv = util::get_domain_attrs(&domain, "group_gen")
-        .inverse()
-        .unwrap();
+    // FIXME: Should we just call `domain.group_gen_inv()` here?
```
Yes, that's better

CPerezz replied:
I don't agree. Until the PR we made to arkworks does not land, I'd not mix the two ways of getting data from the struct TBH. I know the `util::` is ugly, but if we have the two ways at the same time makes it really confusing.

davidnevadoc replied:
In any case, we should use `group_gen_inv` and remove the `inverse()` call

davidnevadoc reviewed (commented):
Now that we are using `domain.group_gen()` it would be good to use it in place of `domain.element(1)` ( [proof.rs/L247](https://github.com/rust-zkp/ark-plonk/blob/a8110966b777a2eaf40556587814ec9189598ab0/src/proof_system/proof.rs#L247)) These calls return the same element . However I think using the first option is clearer and more consistent.

CPerezz commented on .github/workflows/ci.yml:
```diff
@@ -76,7 +80,7 @@ jobs:
         with:
           command: doc
-          args: --all --document-private-items
+          args: --workspace --document-private-items --all-features
```
Not sure why we need `workspace` when this is a single crate. Could you elaborate on that?

bhgomes replied:
`--all` is the deprecated form of `--workspace` so I was just bringing it up to date.

Also, note that I was able to push changes to the CI which is not something non-contributors should be allowed to do. You should probably change the permissions settings so that only core contributors can change actions files.

CPerezz commented on src/circuit.rs:
```diff
@@ -293,13 +291,16 @@ where
+
     /// Compiles the circuit by using a function that returns a `Result`
     /// with the `ProverKey`, `VerifierKey` and the circuit size.
+    #[allow(clippy::type_complexity)] // NOTE: Clippy is too hash here.
```
We can maybe move this to the `lib.rs` if we thing will pop up more.

bhgomes replied:
NB: I meant `harsh` there instead of `hash`.

It's generally not a good idea to dismiss `clippy` lints globally unless they are really broken or too strict like `pedantic`. Lints related to code complexity are important for constraining the design space and sometimes `clippy` is "wrong" in that it treats all complexity the same using a length heuristic, when that's not always what makes something complex or not. See for example other parts of the library where `FIXME` was the note instead for complexity-related lints. It's important to distinguish these two cases when we know the readability and design are not ideal, and when `clippy` is just not subtle enough.

CPerezz commented on src/constraint_system/composer.rs:
```diff
@@ -27,6 +27,9 @@ use hashbrown::HashMap;
 use num_traits::{One, Zero};
 use std::collections::BTreeMap;
 
+#[cfg(feature = "trace")]
```
Why do we want this feature? And also, in case is needed? Shouldn't we enable it at `Cargo.toml` level with a feature-flag like it's done for `parallel` and `asm`?

bhgomes replied:
This is there because those are only imported when `trace` is enabled. I think the `Cargo.toml` is a bit strange here with the `trace` flag and the `trace-print` flag. Usually, you want features to group dependencies or large sections of a library but in this case it seems to be that `trace` is used for only one function and doesn't relate to any dependencies. Maybe this function `check_circuit_satisfied` makes sense to return a boolean, but certainly, it is a little confusing. I hadn't seen this in my code review.

CPerezz replied:
Ohh that was my bad. I missread and thought BigNumber was added as dep. And therefore wanted to constrain it's usage via features in `Cargo.toml` directly.

CPerezz commented on src/constraint_system/ecc/scalar_mul/fixed_base.rs:
```diff
@@ -28,20 +28,21 @@ where
+    // This function is optimized for fixed base ops **ONLY** and therefore,
+    // the **ONLY** `generator` inputs that should be passed to this
+    // function as inputs are [`dusk_jubjub::GENERATOR`] or
+    // [`dusk_jubjub::GENERATOR_NUMS`].
```
Edit the coment and fix it directly instead of leaving the FIXME please. :)

bhgomes replied:
I wrote this because I didn't know what the new comment should be.

CPerezz replied:
Something like:
```
    // # Note
    // This function is optimized for fixed base ops **ONLY** and therefore,
    // the **ONLY** input that should be passed to the function as a point is the
    // generator or basepoint of the curve over which we are operating.
```
should do. :smile:

CPerezz commented on src/constraint_system/ecc/scalar_mul/fixed_base.rs:
```diff
@@ -313,6 +313,7 @@ mod tests {
         assert!(res.is_ok());
     }
 
+    #[allow(non_snake_case)] // NOTE: We want to use capital `G` and `H`.
```
We should move this lint to `lib.rs` and allow it for the entire lib. As sometimes we want to use capital letter variables.

bhgomes replied:
Honestly, I'm of the opinion you should not use non-Rust conventions for naming and it should only be used in cases where you are trying to match some specification in which case it's more important to get the spec and impl as close to each other as possible. Like code complexity lints, it should be disabled in a more precise way than just everywhere. I just did the smallest change here to remove the issue.

CPerezz replied:
So while I understand your POV, and i respect it, with EC libs, sometimes it's much much easier to read code when you can identify variables and differenciate scalars and points just by seeing whether they're capital or non-capital letters. 

Let's keep the lint for now. But I'm slightly biased on this one to move the lint to be general for the entire library..

CPerezz commented on src/constraint_system/ecc/scalar_mul/fixed_base.rs:
```diff
@@ -379,6 +380,7 @@ mod tests {
         assert!(res.is_ok());
     }
 
+    #[allow(non_snake_case)] // NOTE: We want to use capital `P`.
```
ditto.

CPerezz commented on src/error.rs:
```diff
@@ -76,6 +76,11 @@ pub enum Error {
+// NOTE: This type is not very useful in this crate until this PR is merged:
+// <https://github.com/arkworks-rs/algebra/pull/350>.
+pub type Result<T = ()> = core::result::Result<T, Error>;
+
```
Shouldn't we then leave an issue waiting for the PR instead of adding code that we don't yet need directly?

bhgomes replied:
Sorry, that was an oversight. I meant to delete that before pushing.

CPerezz replied:
Can you then delete it please? Or you left it due to any other reason? :)

bhgomes replied:
I was just waiting for your comments to do a commit.

CPerezz commented on src/permutation/constants.rs:
```diff
@@ -1,20 +1,21 @@
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
+#![allow(non_snake_case)]
```
Remove this from here and just add it in `lib.rs` to avoid repetition.

bhgomes replied:
Disabling lints in a precise way is not repetition, as each one should be used to indicate some important convention in that region of code. Naming lints can also be used as compile-time sanity checking.

CPerezz commented on src/permutation/permutation.rs:
```diff
@@ -966,7 +964,7 @@ mod test {
 
         let domain =
             GeneralEvaluationDomain::<E::Fr>::new(num_wire_mappings).unwrap();
-        let w = util::get_domain_attrs(&domain, "group_gen");
+        let w = domain.group_gen();
```
I would prefer to avoid touching anything related to domain internal field obtentions until https://github.com/arkworks-rs/algebra/pull/333 is released.

This will make it much more easy and organic. But for now I would prefer to get all of the elements on the same way. Instead of some on one way, and some on the other..

bhgomes replied:
Not sure what you mean by "some one way and some on the other". This PR removed the entire `util::get_domain_attrs` method.

CPerezz commented on src/permutation/permutation.rs:
```diff
@@ -1105,39 +1103,39 @@ mod test {
         */
         let domain =
             GeneralEvaluationDomain::<E::Fr>::new(num_wire_mappings).unwrap();
-        let w = util::get_domain_attrs(&domain, "group_gen");
+        let w = domain.group_gen();
```
ditto.

CPerezz commented on src/permutation/permutation.rs:
```diff
@@ -1266,8 +1264,7 @@ mod test {
         // Check that z(w^{n+1}) == z(1) == 1
         // This is the first check in the protocol
         assert_eq!(z_poly.evaluate(&F::one()), F::one());
-        let n_plus_one = domain.elements().last().unwrap()
-            * &util::get_domain_attrs(&domain, "group_gen");
```
ditto.

CPerezz commented on src/proof_system/linearisation_poly.rs:
```diff
@@ -74,6 +74,7 @@ pub(crate) struct ProofEvaluations<F: PrimeField> {
 }
 
 /// Compute the linearisation polynomial.
+#[allow(clippy::type_complexity)] // FIXME: Make a separate struct for the tuple arguments.
```
ditto.

bhgomes replied:
Note how this lint indicates a change to be made rather than an oversight on the linter. Those two kinds of situations are important to distinguish between. See my comments above about linting in general.

CPerezz replied:
I understand your point. But I do also consider the priorities of the lib.

I agree, if we can afford reducing the complexity of the types and fix this would be nice. But that should never be in charge of increasing the complexity of the implementation complexity itself. As the algorithm is already complex and difficult sometimes to match to the paper. So if to satisfy clippy and some Rust standards we need to deviate from the implementation style a lot, I would not feel it's worth TBH.

That was my concern when I pointed the clippy lint complaints.

CPerezz commented on src/proof_system/preprocess.rs:
```diff
@@ -222,6 +224,7 @@ impl<E: PairingEngine, P: TEModelParameters<BaseField = E::Fr>>
     /// must perform IFFTs on the selector polynomials and permutation
     /// polynomials in order to commit to them and have the same transcript
     /// view.
+    #[allow(clippy::type_complexity)] // FIXME: Add struct for preprocessed key.
```
ditto.

bhgomes replied:
Note how this lint indicates a change to be made rather than an oversight on the linter. Those two kinds of situations are important to distinguish between. See my comments above about linting in general.

CPerezz replied:
Ditto response above.

CPerezz commented on src/proof_system/proof.rs:
```diff
@@ -479,9 +481,9 @@ fn compute_barycentric_eval<F: PrimeField>(
-        * util::get_domain_attrs(&domain, "size_inv");
-    let range = (0..evaluations.len()).into_iter();
+    let numerator =
+        domain.evaluate_vanishing_polynomial(point) * domain.size_inv();
```
ditto for `domain.size_inv()`

CPerezz commented on src/proof_system/prover.rs:
```diff
@@ -105,6 +92,7 @@ impl<
     }
 
     /// Split `t(X)` poly into 4 degree `n` polynomials.
+    #[allow(clippy::type_complexity)] // NOTE: This is an ok type for internal use.
```
ditto.

bhgomes replied:
See above.

CPerezz commented on src/util.rs:
```diff
@@ -52,34 +52,77 @@ pub fn ruffini<F: PrimeField>(
+    #[inline]
+    fn log_size_of_group(&self) -> u32 {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.log_size_of_group,
+            GeneralEvaluationDomain::MixedRadix(domain) => {
+                domain.log_size_of_group
+            }
+        }
+    }
+
+    #[inline]
+    fn size_inv(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.size_inv,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.size_inv,
+        }
+    }
+
+    #[inline]
+    fn group_gen(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.group_gen,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.group_gen,
+        }
+    }
+
+    #[inline]
+    fn group_gen_inv(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.group_gen_inv,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.group_gen_inv,
+        }
+    }
+
+    #[inline]
+    fn generator_inv(&self) -> F {
+        match self {
+            GeneralEvaluationDomain::Radix2(domain) => domain.generator_inv,
+            GeneralEvaluationDomain::MixedRadix(domain) => domain.generator_inv,
+        }
```
That is basically replaced by https://github.com/arkworks-rs/algebra/pull/333/files as soon as is released. So I won't merge it here. Specially considering the repo should ideally contain only PLONK-related logic and avoid adding any external stuff as much as possible (the solution is really good obviously) But I prefer to wait for the new version rather than have a lot of new code that we will need to remove soon anyway.

bhgomes replied:
Here are some reasons why this should be kept:
1. This repo is already containing non-PLONK logic. This is the nature of software and we cannot expect our code to be a spec implementation only.
2. Using `&'static str` to query members of an object cannot be checked by the compiler and so when we do something wrong like mistype the name we cannot know if it is wrong unless we have a test that passes through that line and even then you might not have a good way of finding that bug. Getting full code coverage is harder than just using extension `trait`s.
3. Using extension traits to extend non-local libraries is standard practice in Rust so other developers will recognize this design language and understand its purpose more readily.
4. Using a trait allows for better discoverability and documentation.
5. Using a trait allows for abstraction over implementations.
6. If/when `arkworks` decides to release the new version, the change here is just to delete the extension trait, everything else stays the same.

CPerezz replied:
After discussing with @davidnevadoc he agreed on leaving the struct/trait that you purposed. 

So fine from my side then.

CPerezz reviewed (changes requested):
Thanks for the PR!

Two main things.
- I would delete everything related to `EvaluationDomainExt`. You're trying to solve in this crate a design issue of a different crate. Which has already been addressed and is waiting for a release. See: https://github.com/arkworks-rs/algebra/pull/333 (It's basically the same idea you coded :smile: )
- I'm not a fan of the removal of the `PlonkVerifierKey` alias. And the reason is that in `ark-poly-commit` there are quite some structs/traits called on the same way. And in advance to that, we changed the name of our own structure to make the distinction of what's PLONK-related and what's polynomial-commitment related. So personally, I would prefer to un-do it.

Other than that a few nits and suggestions. But thanks a lot for your contribution @bhgomes  we really appreciate it! Sorry for the comments and change requests...

CPerezz reviewed (approved):
Approved modulo de removal of the `Result` type inclusion.

LukePearson1 reviewed (approved):
Top code quality @bhgomes, I agree with the trait abstractions. LGTM overall.

On a side note: good to see more people getting involved in open source work.

bhgomes committed:
fix: resolve comments, suggestions, and some formatting

- fixed some `clippy` lints directly instead of leaving `allow` cues, left others for future refactoring
- fixed old jubjub comment
- removed `crate::error::Result` type
- remove some type declaration redundancies
- sorted `derive` macro invocations
- added warning denial to CI
- resolved other PR comments

davidnevadoc reviewed (approved):
LGTM 👍
